<a href="https://colab.research.google.com/github/dxmai/CS114.L21.KHCL/blob/main/FinalProject/FakeNews_Data_Include_Text.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import json

#Trang tin giả

##Activistpost

In [ ]:
import scrapy

class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://www.activistpost.com/'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        getLinkInDiv = response.css('h3.content-list-title')
        allLinks = getLinkInDiv.css('a::attr(href)').getall()

        for each_link in allLinks:
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('a.next::attr(href)').get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)


    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018'] 
        available = False 
        date = response.css('.entry-meta span::text').get() 
        for year in yearsList: 
            if year in date: 
                available = True 
                break
        if available: 
            title = response.css('.entry-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.entry-content p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 4:
                    break
                if count == 0:
                    count += 1
                    continue
                sentence = sentence.replace('\xa0', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }

In [24]:
activist = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/activistpost.json')
activist.dropna(inplace=True)
activist

date  ... is_fake
0     2021-05-28  ...       1
1     2021-06-28  ...       1
2     2021-05-29  ...       1
3     2021-05-28  ...       1
4     2021-06-25  ...       1
...          ...  ...     ...
10270 2019-12-06  ...       1
10271 2019-12-06  ...       1
10272 2019-12-05  ...       1
10273 2019-12-05  ...       1
10274 2019-12-05  ...       1

[10275 rows x 4 columns]

##Natural News

In [ ]:
import scrapy


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://www.naturalnews.com/all-posts.html'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        getLinkInDiv = response.css('div.f-tabbed-list-content')
        allLinks = getLinkInDiv.css('h2 a::attr(href)').getall()

        for each_link in allLinks:
            each_link = 'https://www.naturalnews.com/' + each_link
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('.pagination-next a::attr(href)').get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        date = response.css('.Article-Author::text').get()
        date = date.replace(' by: ', '')
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('.entry-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.entry-content p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 4:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }

In [23]:
natural = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/naturalnews.json')
natural.dropna(inplace=True)
natural

date  ... is_fake
0     2021-06-29  ...       1
1     2021-06-29  ...       1
2     2021-06-29  ...       1
3     2021-06-29  ...       1
4     2021-06-29  ...       1
...          ...  ...     ...
11980 2020-03-03  ...       1
11981 2020-03-03  ...       1
11982 2020-03-03  ...       1
11983 2020-03-04  ...       1
11984 2020-03-05  ...       1

[11985 rows x 4 columns]

##Zerohedge

In [ ]:
import scrapy
dateValid = True

class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://www.zerohedge.com/'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        getLinkInDiv = response.css('div.Article_mobileNonSticky__PmGNH')
        allLinks = getLinkInDiv.css('h2 a::attr(href)').getall()

        for each_link in allLinks:
            if dateValid:
                each_link = 'https://www.zerohedge.com' + each_link
                yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('a.SimplePaginator_next__15okP::attr(href)').get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        global dateValid
        date = response.css('.ArticleFull_headerFooter__date__3T7FN::text').get()
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('.ArticleFull_title__2cUI6::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.NodeContent_body__2clki p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 4:
                    break
                sentence = sentence.replace('\xa0', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }
        else:
            dateValid = False

In [22]:
zerohedge = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/zerohedge.json')
zerohedge.dropna(inplace = True)
zerohedge

date  ... is_fake
0   2021-07-01 09:05:00  ...       1
1   2021-07-01 09:28:00  ...       1
2   2021-07-01 08:35:00  ...       1
3   2021-07-01 08:45:00  ...       1
4   2021-07-01 07:15:00  ...       1
..                  ...  ...     ...
366 2021-06-30 20:05:00  ...       1
367 2021-07-01 12:26:00  ...       1
368 2021-07-01 12:30:00  ...       1
369 2021-07-01 06:44:00  ...       1
370 2021-07-01 12:41:00  ...       1

[371 rows x 4 columns]

##Prntly

In [ ]:
import scrapy
dateValid = True

class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://prntly.com/category/science/',
            'https://prntly.com/category/politics/',
            'https://prntly.com/category/world/europe/',
            'https://prntly.com/category/world/asia/',
            'https://prntly.com/category/world/middle-east/',
            'https://prntly.com/category/local-news/',
            'https://prntly.com/category/trade-jobs/',
            'https://prntly.com/category/immigration/'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        getLinkInDiv = response.css('div.entry-content-holder')
        allLinks = getLinkInDiv.css('h2 a::attr(href)').getall()

        for each_link in allLinks:
            if dateValid:
                yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('a.next::attr(href)').get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        global dateValid
        date = response.css('.entry-date::text').get()
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('h1.entry-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.entry-content p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 4:
                    break
                sentence = sentence.replace('\xa0', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }
        else:
            dateValid = False


In [21]:
prntly = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/prntly.json')
prntly.dropna(inplace=True)
prntly

date  ... is_fake
0   2020-01-28  ...       1
1   2020-02-07  ...       1
2   2020-01-27  ...       1
3   2020-02-28  ...       1
4   2020-02-28  ...       1
..         ...  ...     ...
568 2020-01-13  ...       1
569 2020-01-13  ...       1
570 2020-01-13  ...       1
571 2020-01-13  ...       1
572 2020-01-13  ...       1

[573 rows x 4 columns]

##Thegateway

In [ ]:
import scrapy
dateValid = True

class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://www.thegatewaypundit.com/'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        if dateValid:
            getLinkInDiv = response.css('div.tgp-post')
            allLinks = getLinkInDiv.css('.entry-archive-title a::attr(href)').getall()

        for each_link in allLinks:
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('.next a::attr(href)').get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        global dateValid
        date = response.css('.entry-meta-text span::text')[2].get()
        date = date.replace('\nPublished ', '')
        temp = date.split(' at', 1)
        date = temp[0]
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('h1.entry-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.entry-content p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 3:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }
        else:
            dateValid = False


In [20]:
thegateway = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/thegateway.json')
thegateway.dropna(inplace=True)
thegateway

date  ... is_fake
0       July 1, 2021  ...       1
1       July 1, 2021  ...       1
2       July 1, 2021  ...       1
3       July 1, 2021  ...       1
4       July 1, 2021  ...       1
...              ...  ...     ...
36658  June 30, 2021  ...       1
36659  June 30, 2021  ...       1
36660  June 30, 2021  ...       1
36661  June 30, 2021  ...       1
36662   July 1, 2021  ...       1

[36663 rows x 4 columns]

##Conservativedailypost 

In [ ]:
import scrapy
dateValid = True


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://conservativedailypost.com/category/politics/'
            'https://conservativedailypost.com/category/u-s/',
            'https://conservativedailypost.com/category/world/'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        if dateValid:
            getLinkInDiv = response.css('div.mob')
            allLinks = getLinkInDiv.css('h4 a::attr(href)').getall()

        for each_link in allLinks:
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('a.next::attr(href)').get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        global dateValid
        date = response.css('.date-published::text')[-1].get()
        date = date.replace('\n', '')
        date = date.replace('\t', '')
        date = date.replace('\r', '')
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('h1.min-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.content-container p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 3:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\r', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }
        else:
            dateValid = False

In [19]:
conservative = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/conservative.json')
conservative.dropna(inplace=True)
conservative

date  ... is_fake
0    2021-07-01 13:00:00  ...       1
1    2021-07-01 17:54:00  ...       1
2    2021-07-01 18:18:00  ...       1
3    2021-07-01 19:01:00  ...       1
4    2021-07-01 16:41:00  ...       1
...                  ...  ...     ...
5121 2018-01-23 00:29:00  ...       1
5122 2018-01-23 00:23:00  ...       1
5123 2018-01-23 00:16:00  ...       1
5124 2018-01-22 23:45:00  ...       1
5125 2018-01-21 00:52:00  ...       1

[5126 rows x 4 columns]

##Dailyheadlines

In [ ]:
import scrapy
dateValid = True


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'http://dailyheadlines.com/'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        if dateValid:
            getLinkInDiv = response.css('h2.entry-title')
            allLinks = getLinkInDiv.css('a::attr(href)').getall()

        for each_link in allLinks:
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('.pagination a::attr(href)')[-1].get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        global dateValid
        date = response.css('.post-meta-date::text')[-1].get()
        date = date.replace('\n', '')
        date = date.replace(' ' * 21, '')
        print(date)
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('h2.post-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            title = title.replace(' ' * 12, '')
            title = title.replace(' ' * 9, '')
            paragraph = response.css('.entry-content p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 3:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\r', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 1
            }
        else:
            dateValid = False



In [18]:
dailyheadline = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/dailyheadline.json')
dailyheadline.dropna(inplace=True)
dailyheadline

date  ... is_fake
0    2021-06-25  ...       1
1    2021-06-25  ...       1
2    2021-06-24  ...       1
3    2021-07-01  ...       1
4    2021-06-25  ...       1
...         ...  ...     ...
2573 2021-04-08  ...       1
2574 2021-05-19  ...       1
2575 2021-06-24  ...       1
2576 2021-06-23  ...       1
2577 2021-06-24  ...       1

[2578 rows x 4 columns]

#Trang tin chính thống

##Csmonitor **Bị chặn tên miền rồiiiiiii** ;;-;;

In [ ]:
import scrapy
dateValid = True


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://www.csmonitor.com/World/Americas/(view)/all',
            'https://www.csmonitor.com/World/Making-a-difference/(view)/all',
            'https://www.csmonitor.com/World/Points-of-Progress/(view)/all',
            'https://www.csmonitor.com/World/(view)/all',
            'https://www.csmonitor.com/World/Africa/(view)/all',
            'https://www.csmonitor.com/World/Asia-Pacific/(view)/all',
            'https://www.csmonitor.com/World/Asia-South-Central/(view)/all',
            'https://www.csmonitor.com/World/Europe/(view)/all',
            'https://www.csmonitor.com/World/Middle-East/(view)/all',
            'https://www.csmonitor.com/USA/(view)/all',
            'https://www.csmonitor.com/USA/Politics/(view)/all',
            'https://www.csmonitor.com/USA/Politics/monitor_breakfast/(view)/all',
            'https://www.csmonitor.com/USA/Politics/Politics-Watch/(view)/all',
            'https://www.csmonitor.com/USA/Foreign-Policy/(view)/all',
            'https://www.csmonitor.com/USA/Military/(view)/all',
            'https://www.csmonitor.com/USA/Justice/(view)/all',
            'https://www.csmonitor.com/USA/Society/(view)/all',
            'https://www.csmonitor.com/Business/(view)/all',
            'https://www.csmonitor.com/Science/(view)/all',
            'https://www.csmonitor.com/Environment/(view)/all',
            'https://www.csmonitor.com/Technology/(view)/all',
            'https://www.csmonitor.com/The-Culture/(view)/all',
            'https://www.csmonitor.com/The-Culture/Arts/(view)/all',
            'https://www.csmonitor.com/The-Culture/Music/(view)/all',
            'https://www.csmonitor.com/The-Culture/Movies/(view)/all',
            'https://www.csmonitor.com/The-Culture/TV/(view)/all',
            'https://www.csmonitor.com/The-Culture/In-a-Word/(view)/all',
            'https://www.csmonitor.com/The-Culture/The-Home-Forum/(view)/all',
            'https://www.csmonitor.com/The-Culture/Food/(view)/all'
        ]
        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        if dateValid:
            getLinkInDiv = response.css('div.ezv-listing')
            allLinks = getLinkInDiv.css('a::attr(href)').getall()

        for each_link in allLinks:
            each_link = 'https://www.csmonitor.com' + each_link
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = 'https://www.csmonitor.com' + response.css('.ui-brdr-widget a::attr(href)')[-1].get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        global dateValid

        date = response.css('.eza-publish_date::text').get()
        date = date.replace('\n', '')
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('h1.eza-title::text').get()
            title = title.replace('\n', '')
            title = title.replace('\t', '')
            paragraph = response.css('.story-two p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 3:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\r', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 0
            }
        else:
            dateValid = False

##Economist

In [ ]:
import scrapy


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://www.economist.com/leaders',
            'https://www.economist.com/briefing',
            'https://www.economist.com/united-states',
            'https://www.economist.com/asia',
            'https://www.economist.com/the-americas',
            'https://www.economist.com/china',
            'https://www.economist.com/middle-east-and-africa',
            'https://www.economist.com/europe',
            'https://www.economist.com/britain',
            'https://www.economist.com/international',
            'https://www.economist.com/business',
            'https://www.economist.com/finance-and-economics',
            'https://www.economist.com/science-and-technology',
            'https://www.economist.com/books-and-arts',
            'https://www.economist.com/graphic-detail',
            'https://www.economist.com/obituary'
        ]

        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        getLinkInDiv = response.css('div.teaser__text')
        allLinks = getLinkInDiv.css('a::attr(href)').getall()

        for each_link in allLinks:
            each_link = 'https://www.economist.com' + each_link
            yield scrapy.Request(each_link, callback=self.parse_href)

        next_page = response.css('.ds-pagination__nav--next a::attr(href)')[0].get()
        if next_page is not None:
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        date = response.css('.article__dateline-datetime::text').get()
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('.article__headline::text').get()
            paragraph = response.css('.article__body-text::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 3:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\r', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 0
            }

In [17]:
eco = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/economist.json')
eco.dropna(inplace=True)
eco

date  ... is_fake
0     2021-06-24  ...       0
1     2021-06-24  ...       0
2     2021-06-24  ...       0
3     2021-06-24  ...       0
4     2021-06-12  ...       0
...          ...  ...     ...
12600 2020-08-20  ...       0
12601 2019-01-17  ...       0
12602 2018-09-20  ...       0
12603 2018-04-14  ...       0
12604 2018-08-18  ...       0

[12605 rows x 4 columns]

##APNews

In [ ]:
import scrapy


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        urls = [
            'https://apnews.com/hub/coronavirus-pandemic',
            'https://apnews.com/hub/politics',
            'https://apnews.com/hub/sports',
            'https://apnews.com/hub/entertainment',
            'https://apnews.com/hub/lifestyle',
            'https://apnews.com/hub/oddities',
            'https://apnews.com/hub/travel',
            'https://apnews.com/hub/technology',
            'https://apnews.com/hub/ap-fact-check',
            'https://apnews.com/hub/business',
            'https://apnews.com/hub/us-news',
            'https://apnews.com/hub/health',
            'https://apnews.com/hub/science',
            'https://apnews.com/hub/world-news',
            'https://apnews.com/hub/religion'
        ]

        for url in urls:
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        getLinkInDiv = response.css('div.CardHeadline')
        allLinks = getLinkInDiv.css('a::attr(href)').getall()

        for each_link in allLinks:
            each_link = 'https://apnews.com' + each_link
            yield scrapy.Request(each_link, callback=self.parse_href)

        #next_page = response.css('.ds-pagination__nav--next a::attr(href)')[0].get()
        #if next_page is not None:
        #    next_page = response.urljoin(next_page)
        #    yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        date = response.css('.Timestamp::attr(title)').get()
        temp = date.split(' ', 1)
        date = temp[0]
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('.CardHeadline h1::text').get()
            paragraph = response.css('.Article p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 3:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\r', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 0
            }

In [ ]:
apnews = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/apnews.json')
apnews

date  ... is_fake
0   2021-07-02  ...       0
1   2021-07-03  ...       0
2   2021-07-02  ...       0
3   2021-07-02  ...       0
4   2021-05-19  ...       0
..         ...  ...     ...
731 2021-07-01  ...       0
732 2021-07-01  ...       0
733 2021-07-02  ...       0
734 2021-07-03  ...       0
735 2021-07-01  ...       0

[736 rows x 4 columns]

In [ ]:
apnews.dropna(inplace=True)
apnews

date  ... is_fake
0   2021-07-02  ...       0
1   2021-07-03  ...       0
2   2021-07-02  ...       0
3   2021-07-02  ...       0
4   2021-05-19  ...       0
..         ...  ...     ...
731 2021-07-01  ...       0
732 2021-07-01  ...       0
733 2021-07-02  ...       0
734 2021-07-03  ...       0
735 2021-07-01  ...       0

[730 rows x 4 columns]

##Npr (**load_more style**)

In [ ]:
import scrapy
offset, times = 0, 0
valid = True
res = ''
dictionary = {
    'national/': [1003, 25],
    'politics/': [1014, 25],
    'health/': [1128, 23],
    'technology/': [1019, 24],
    'world/': [1004, 22],
    'business/': [1006, 25],
    'science/': [1007, 24]
}


class NewsSpider(scrapy.Spider):
    name = "news"

    def start_requests(self):
        global times, res, offset, valid
        urls = [
            'https://www.npr.org/sections/national/',
            'https://www.npr.org/sections/politics/',
            'https://npr.org/sections/health/'
            'https://npr.org/sections/technology/'
            'https://www.npr.org/sections/world/'
            'https://www.npr.org/sections/business/'
            'https://www.npr.org/sections/science/',
        ]

        for url in urls:
            for key in dictionary.keys():
                if url.endswith(key):
                    res = key
                    break
            times = 0
            offset = 0
            valid = True
            yield scrapy.Request(url, callback=self.parse)

    def parse(self, response):
        global offset, times
        getLinkInDiv = response.css('h2.title')
        allLinks = getLinkInDiv.css('a::attr(href)').getall()

        for each_link in allLinks:
            yield scrapy.Request(each_link, callback=self.parse_href)

        if valid:
            next_page = 'https://www.npr.org/get/'
            temp = dictionary[res][0]
            next_page = next_page + str(temp)
            next_page = next_page + '/render/partial/next?start='
            if times == 0:
                offset += dictionary[res][1]
                times += 1
            else:
                offset += 24
            next_page = next_page + str(offset)
            next_page = response.urljoin(next_page)
            yield scrapy.Request(next_page, callback=self.parse)

    def parse_href(self, response):
        global valid
        yearsList = ['2021', '2020', '2019', '2018']
        available = False
        date = response.css('.date::text').get()
        for year in yearsList:
            if year in date:
                available = True
                break
        if available:
            title = response.css('.storytitle h1::text').get()
            paragraph = response.css('.storylocation p::text').getall()
            text = ''
            count = 0
            for sentence in paragraph:
                if count == 0 or count == 1 or count == 2:
                    count += 1
                    continue
                if count == 6:
                    break
                sentence = sentence.replace('\xa0', '')
                sentence = sentence.replace('\n', '')
                sentence = sentence.replace('\r', '')
                sentence = sentence.replace('\t', '')
                text = text + sentence
                count += 1
            yield {
                'date': date,
                'title': title,
                'text': text,
                'is_fake': 0
            }
        else:
            valid = False

In [10]:
npr_national = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_national.json')
npr_business = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_business.json')
npr_science = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_science.json')
npr_health = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_health.json')
npr_tech = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_tech.json')
npr_world = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_world.json')
npr_politic = pd.read_json('https://raw.githubusercontent.com/dxmai/CS114.L21.KHCL/main/FinalProject/Include_Text/npr_politics.json')

In [15]:
npr_list = [npr_science, npr_health, npr_tech, npr_world, npr_politic]
dataset = npr_national.append(npr_business, ignore_index=True)
for i in npr_list:
  dataset = dataset.append(i, ignore_index=True)
dataset.dropna(inplace=True)
dataset

date  ... is_fake
0      Updated July 2, 2021  ...       0
1      Updated July 2, 2021  ...       0
2      Updated July 1, 2021  ...       0
3      Updated July 1, 2021  ...       0
4              July 1, 2021  ...       0
...                     ...  ...     ...
45299        March 13, 2020  ...       0
45300        March 13, 2020  ...       0
45301        March 12, 2020  ...       0
45302        March 13, 2020  ...       0
45303        March 12, 2020  ...       0

[45290 rows x 4 columns]

#Tổng kết số lượng

Tổng kết số lượng tìm được (đã loại bỏ những trường hợp có giá trị bị null):

Tin thật: 45290 + 730 + 12605 = 58625

Tin giả: 2578 + 5126 + 36663 + 573 + 371 + 11985 +  10275 = 67571